File: src/TaxAuthorityProvider/TaxAuthorities/Slovenia/ReportGeneration/IFI/XML_D_IFI.py
from typing import Sequence

from lxml import etree

import Core.FinancialEvents.Schemas.Grouping as pgf
import TaxAuthorityProvider.Schemas.Configuration as tc
import TaxAuthorityProvider.TaxAuthorities.Slovenia.ReportGeneration.IFI.Common as common
import TaxAuthorityProvider.TaxAuthorities.Slovenia.Schemas.Schemas as ss
from Core.FinancialEvents.Services.FinancialEventsProcessor import (
    FinancialEventsProcessor,
)


def generateXmlReport(
    reportConfig: tc.TaxAuthorityConfiguration,
    data: Sequence[pgf.FinancialGrouping],
    templateEnvelope: etree._Element,
    countedProcessor: FinancialEventsProcessor,
) -> etree._Element:
    convertedTrades = common.convertTradesToIfiItems(reportConfig, data, countedProcessor)

    nsmap = templateEnvelope.nsmap
    nsmap[None] = "http://edavki.durs.si/Documents/Schemas/D_IFI_4.xsd"
    envelope = etree.Element(templateEnvelope.tag, attrib=templateEnvelope.attrib, nsmap=nsmap)
    envelope[:] = templateEnvelope[:]

    body = etree.SubElement(envelope, "body")
    etree.SubElement(body, etree.QName(nsmap["edp"], "bodyContent"))

    D_IFI = etree.SubElement(body, "D_IFI")
    etree.SubElement(D_IFI, "PeriodStart").text = reportConfig.fromDate.format("YYYY-MM-DD")
    etree.SubElement(D_IFI, "PeriodEnd").text = reportConfig.toDate.shift(days=-1).format("YYYY-MM-DD")
    # etree.SubElement(KDVP, "CountryOfResidenceID").text = self.taxPayerInfo.countryID
    # etree.SubElement(D_IFI, "TelephoneNumber").text = self.taxPayerInfo.PhoneNumber
    # etree.SubElement(D_IFI, "Email").text = self.taxPayerInfo.email

    for DIFI_item_entry in convertedTrades:
        DIFI_ITEM = etree.SubElement(D_IFI, "TItem")

        etree.SubElement(DIFI_ITEM, "TypeId").text = DIFI_item_entry.ItemType.value
        etree.SubElement(DIFI_ITEM, "Type").text = DIFI_item_entry.InventoryListType.value

        if DIFI_item_entry.InventoryListType == ss.EDavkiDerivativeSecurityType.OPTION_OR_CERTIFICATE:
            etree.SubElement(DIFI_ITEM, "TypeName").text = ss.EDavkiDerivativeSecurityTypeName.OPTION_OR_CERTIFICATE.value
        elif DIFI_item_entry.InventoryListType == ss.EDavkiDerivativeSecurityType.FUTURES_CONTRACT:
            etree.SubElement(DIFI_ITEM, "TypeName").text = ss.EDavkiDerivativeSecurityTypeName.FUTURES_CONTRACT.value
        elif DIFI_item_entry.InventoryListType == ss.EDavkiDerivativeSecurityType.CONTRACT_FOR_DIFFERENCE:
            etree.SubElement(DIFI_ITEM, "TypeName").text = ss.EDavkiDerivativeSecurityTypeName.CONTRACT_FOR_DIFFERENCE.value
        else:
            etree.SubElement(DIFI_ITEM, "TypeName").text = ss.EDavkiDerivativeSecurityTypeName.OTHER.value

        if DIFI_item_entry.Name is not None:
            etree.SubElement(DIFI_ITEM, "Name").text = DIFI_item_entry.Name

        # Code and ISIN are being left out, as the platform doesn't:
        # - Support multiple entries with the same ISIN
        # - Ticker (Code) cannot be longer than 10 characters
        # if DIFI_item_entry.Code is not None:
        #     etree.SubElement(DIFI_ITEM, "Code").text = DIFI_item_entry.Code
        # if DIFI_item_entry.ISIN is not None and DIFI_item_entry.ISIN != "":  # TODO: Figure out why ISIN is empty sometimes
        #     etree.SubElement(DIFI_ITEM, "ISIN").text = DIFI_item_entry.ISIN

        etree.SubElement(DIFI_ITEM, "HasForeignTax").text = str(DIFI_item_entry.HasForeignTax).lower()
        if DIFI_item_entry.ForeignTax is not None:
            etree.SubElement(DIFI_ITEM, "ForeignTax").text = str(DIFI_item_entry.ForeignTax.__round__(8))

        if DIFI_item_entry.FTCountryID is not None:
            etree.SubElement(DIFI_ITEM, "FTCountryID").text = DIFI_item_entry.FTCountryID
        if DIFI_item_entry.FTCountryName is not None:
            etree.SubElement(DIFI_ITEM, "FTCountryName").text = DIFI_item_entry.FTCountryName

        if DIFI_item_entry.ItemType == ss.EDavkiDerivativeReportItemType.DERIVATIVE:

            for entryLine in DIFI_item_entry.Items:
                entry = etree.SubElement(DIFI_ITEM, "TSubItem")

                if isinstance(
                    entryLine,
                    ss.EDavkiDerivativeReportSecurityLineGenericEventBought,
                ):
                    purchase = etree.SubElement(entry, "Purchase")
                    etree.SubElement(purchase, "F1").text = entryLine.BoughtOn.format("YYYY-MM-DD")
                    etree.SubElement(purchase, "F2").text = entryLine.GainType.value
                    etree.SubElement(purchase, "F3").text = str(entryLine.Quantity.__round__(8))
                    etree.SubElement(purchase, "F4").text = str(entryLine.PricePerUnit.__round__(8))
                    etree.SubElement(purchase, "F9").text = str(entryLine.Leveraged).lower()

                if isinstance(entryLine, ss.EDavkiDerivativeReportSecurityLineGenericEventSold):
                    sale = etree.SubElement(entry, "Sale")
                    etree.SubElement(sale, "F5").text = entryLine.SoldOn.format("YYYY-MM-DD")
                    etree.SubElement(sale, "F6").text = str(-entryLine.Quantity.__round__(8))
                    etree.SubElement(sale, "F7").text = str(entryLine.PricePerUnit.__abs__().__round__(8))

    return envelope

File: lib/src/TaxAuthorities/Slovenia/ReportGeneration/Ifi/XmlDIfi.ts
import type { TaxAuthorityConfiguration } from "@brrr/TaxAuthorities/ConfigurationProvider";
import {
	EDavkiDerivativeReportItemType,
	EDavkiDerivativeReportSecurityLineGenericEventBought,
	EDavkiDerivativeSecurityType,
	EDavkiDerivativeSecurityTypeName,
	type EDavkiGenericDerivativeReportItem,
} from "@brrr/TaxAuthorities/Slovenia/Schemas/Schemas";
import { XMLBuilder } from "fast-xml-parser";

function round(value: number, decimals: number): number {
	const factor = 10 ** decimals;
	return Math.round(value * factor) / factor;
}

export function generateXmlReport(
	reportConfig: TaxAuthorityConfiguration,
	ifiItems: EDavkiGenericDerivativeReportItem[],
): string {
	const edp = "http://edavki.durs.si/Documents/Schemas/EDP-Common-1.xsd";

	const tItemNodes = ifiItems.map((ifiItem) => {
		let typeName: string;
		if (ifiItem.inventoryListType === EDavkiDerivativeSecurityType.OPTION_OR_CERTIFICATE) {
			typeName = EDavkiDerivativeSecurityTypeName.OPTION_OR_CERTIFICATE;
		} else if (ifiItem.inventoryListType === EDavkiDerivativeSecurityType.FUTURES_CONTRACT) {
			typeName = EDavkiDerivativeSecurityTypeName.FUTURES_CONTRACT;
		} else if (ifiItem.inventoryListType === EDavkiDerivativeSecurityType.CONTRACT_FOR_DIFFERENCE) {
			typeName = EDavkiDerivativeSecurityTypeName.CONTRACT_FOR_DIFFERENCE;
		} else {
			typeName = EDavkiDerivativeSecurityTypeName.OTHER;
		}

		const tItemObj: Record<string, unknown> = {
			TypeId: ifiItem.itemType,
			Type: ifiItem.inventoryListType,
			TypeName: typeName,
		};

		if (ifiItem.name !== null) {
			tItemObj.Name = ifiItem.name;
		}

		tItemObj.HasForeignTax = String(ifiItem.hasForeignTax).toLowerCase();

		if (ifiItem.foreignTax !== null) {
			tItemObj.ForeignTax = String(round(ifiItem.foreignTax, 8));
		}
		if (ifiItem.ftCountryId !== null) {
			tItemObj.FTCountryID = ifiItem.ftCountryId;
		}
		if (ifiItem.ftCountryName !== null) {
			tItemObj.FTCountryName = ifiItem.ftCountryName;
		}

		if (ifiItem.itemType === EDavkiDerivativeReportItemType.DERIVATIVE) {
			const tSubItemNodes = ifiItem.items.map((entryLine) => {
				if (entryLine instanceof EDavkiDerivativeReportSecurityLineGenericEventBought) {
					return {
						Purchase: {
							F1: entryLine.boughtOn.toFormat("yyyy-MM-dd"),
							F2: entryLine.gainType,
							F3: String(round(entryLine.quantity, 8)),
							F4: String(round(entryLine.pricePerUnit, 8)),
							F9: String(entryLine.leveraged).toLowerCase(),
						},
					};
				} else {
					return {
						Sale: {
							F5: entryLine.soldOn.toFormat("yyyy-MM-dd"),
							F6: String(round(-entryLine.quantity, 8)),
							F7: String(round(Math.abs(entryLine.pricePerUnit), 8)),
						},
					};
				}
			});
			tItemObj.TSubItem = tSubItemNodes;
		}

		return tItemObj;
	});

	const tree = {
		"?xml": { "@_version": "1.0", "@_encoding": "UTF-8" },
		Envelope: {
			"@_xmlns:edp": edp,
			"@_xmlns:xs": "http://www.w3.org/2001/XMLSchema",
			"edp:Header": {
				"edp:taxpayer": {},
				"edp:Workflow": {
					"edp:DocumentWorkflowID": "O",
				},
			},
			"edp:AttachmentList": {},
			"edp:Signatures": {},
			body: {
				"edp:bodyContent": {},
				D_IFI: {
					PeriodStart: reportConfig.fromDate.toFormat("yyyy-MM-dd"),
					PeriodEnd: reportConfig.toDate.minus({ days: 1 }).toFormat("yyyy-MM-dd"),
					TItem: tItemNodes,
				},
			},
		},
	};

	const builder = new XMLBuilder({
		ignoreAttributes: false,
		attributeNamePrefix: "@_",
		format: true,
		suppressEmptyNode: false,
	});

	return builder.build(tree) as string;
}
